crates/edr_evm/src/blockchain/forked.rs

```diff
@@ -57,7 +57,8 @@ pub struct ForkedBlockchain {
     remote: RemoteBlockchain<Arc<dyn SyncBlock<Error = BlockchainError>>, true>,
     state_root_generator: Arc<Mutex<RandomHashGenerator>>,
     fork_block_number: u64,
-    chain_id: u64,
+    chain_id_override: Option<u64>,
```

Comment:
This change is being done because in Hardhat tests we are comparing the returned value of `chainId` opcode with the provided chain ID for the test providers, instead of the actual chain ID of a remote provider (for forks).

I considered this to be a bug in the Hardhat tests, as the `chain_id` opcode  in `revm` should be working with the actual chain ID as reported by the remote provider, for forks.

I fixed it as follows: https://github.com/NomicFoundation/hardhat/pull/4622/files#r1416539300

Feedback from @alcuadrado & @fvictorio would be appreciated.

Reply:
I discussed with Pato and the approach you took is correct. In HH v2 it should be possible to override the `chain_id`, originally designed for security purposes. This behaviour is subject to change in HH v3.

Reply:
One improvement that you can still do is to only store the result of `chain_id_override.unwrap_or(chain_id)`. There's no need to keep both.